bigtable/client/mutations.h

```diff
@@ -131,6 +143,61 @@ class SingleRowMutation {
   google::protobuf::RepeatedPtrField<google::bigtable::v2::Mutation> ops_;
 };
 
+/**
+ * A SingleRowMutation that failed.
+ *
+ * A multi-row mutation returns the list of operations that failed,
+ * this class encapsulates both the failure and the original
+ * mutation.  The application can then choose to resend the mutation,
+ * or log it, or save it for processing via some other means.
+ */
+class FailedMutation {
+ public:
+  FailedMutation(SingleRowMutation mut, google::rpc::Status status)
+      : FailedMutation(std::move(mut), std::move(status), -1) {}
+  FailedMutation(SingleRowMutation mut, google::rpc::Status status, int index)
+      : mutation_(std::move(mut)),
+        status_(to_grpc_status(status)),
+        original_index_(index) {}
+
+  FailedMutation(FailedMutation &&) = default;
+  FailedMutation &operator=(FailedMutation &&) = default;
+
+  //@{
+  /// @name accessors
+  SingleRowMutation const &mutation() const { return mutation_; }
+  grpc::Status const &status() const { return status_; }
+  int original_index() const { return original_index_; }
+  //@}
+
+  friend class MultipleRowMutations;
+
+ private:
+  static grpc::Status to_grpc_status(google::rpc::Status const &status);
+
+ private:
+  SingleRowMutation mutation_;
+  grpc::Status status_;
+  int original_index_;
+};
+
+/**
+ * Report unrecoverable errors in a partially completed mutation.
+ */
+class PermanentMutationFailures : public std::runtime_error {
```

Comment:
I understand that there may be multiple failures that this represents, but it's still weird to see a class named in the plural; I'm used to classes being a singular, but it can contain a list of N failures that happened.

What do you think about making this class name singular?

Reply:
Good point, thanks!